agent/src/main/java/org/openremote/agent/protocol/lorawan/tts/TheThingsStackProtocol.java

```diff
@@ -219,95 +311,712 @@ protected boolean configureMQTTWriteValueTemplate(Attribute<?> attribute, MQTTAg
-                        response.statusCode() + "]: " + response.body()
-                    );
-                    break;
+    private ManagedChannel createChannel(String host, int port, boolean isSecure, boolean withKeepAlive) {
+        ManagedChannelBuilder<?> builder;
+
+        if (isSecure) {
+            builder = ManagedChannelBuilder.forAddress(host, port)
+                .useTransportSecurity();
+        } else {
+            builder =ManagedChannelBuilder.forAddress(host, port)
+                .usePlaintext();
+        }
+
+        if (withKeepAlive) {
+            builder
+                .keepAliveTime(GRPC_BULK_SYNC_RUNNER_KEEP_ALIVE_MILLIS, TimeUnit.MILLISECONDS)
+                .keepAliveTimeout(GRPC_BULK_SYNC_RUNNER_KEEP_ALIVE_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS)
+                .keepAliveWithoutCalls(true);
+        }
+
+        return builder.build();
+    }
+
+    private void performDevicesUpdate(ManagedChannel channel) throws Exception {
+        Optional<String> apiKey = getAgent().getApiKey().map(String::trim);
+        Optional<String> applicationId = getAgent().getApplicationId().map(String::trim);
+
+        if (apiKey.isEmpty() || applicationId.isEmpty()) {
+            throw new IllegalStateException("Missing API key or application ID");
+        }
+
+        Metadata headers = new Metadata();
+        Metadata.Key<String> AUTHORIZATION_HEADER =
+            Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER);
+        headers.put(AUTHORIZATION_HEADER, "Bearer " + apiKey.get());
+
+        EventsGrpc.EventsBlockingStub stub = EventsGrpc.newBlockingStub(channel)
+            .withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers));
+
+
+        EventsOuterClass.StreamEventsRequest request = EventsOuterClass.StreamEventsRequest.newBuilder()
+            .addIdentifiers(
+                Identifiers.EntityIdentifiers.newBuilder()
+                    .setApplicationIds(
+                        Identifiers.ApplicationIdentifiers.newBuilder()
+                            .setApplicationId(applicationId.get())
+                            .build()
+                    )
+            )
+            .build();
+
+        boolean isStreamEstablished = false;
+
+        Iterator<EventsOuterClass.Event> events = stub.stream(request);
+
+        try {
+            while (events.hasNext()) {
+                if (Thread.currentThread().isInterrupted()) {
+                    throw new InterruptedException("[" + DEVICE_EVENT_STREAMER_NAME + "] interrupted" );
                 }
 
-                JsonNode root = mapper.readTree(response.body());
-                JsonNode devices = root.path("end_devices");
+                EventsOuterClass.Event event = events.next();
+
+                isStreamEstablished = true;
+
+                String eventName = event.getName();
+                Any data = event.getData();
+                String typeUrl = data != null ? data.getTypeUrl() : "";
```

Comment:
`data` never seems to be null
```suggestion
                String typeUrl = data.getTypeUrl();
```